components/script/dom/document.rs

```diff
@@ -515,18 +515,38 @@ impl Document {
     /// Attempt to find a named element in this page's document.
     /// https://html.spec.whatwg.org/multipage/#the-indicated-part-of-the-document
     pub fn find_fragment_node(&self, fragid: &str) -> Option<Root<Element>> {
-        self.get_element_by_id(&Atom::from(fragid)).or_else(|| {
-            let check_anchor = |node: &HTMLAnchorElement| {
-                let elem = node.upcast::<Element>();
-                elem.get_attribute(&ns!(), &atom!("name"))
-                    .map_or(false, |attr| &**attr.value() == fragid)
-            };
-            let doc_node = self.upcast::<Node>();
-            doc_node.traverse_preorder()
-                    .filter_map(Root::downcast)
-                    .find(|node| check_anchor(&node))
-                    .map(Root::upcast)
-        })
+        // Step 2
+        if fragid == "" {
+            self.GetDocumentElement()
+        } else {
+            use url::percent_encoding::percent_decode;
+            // Step 3 & 4
+            String::from_utf8(percent_decode(fragid.as_bytes())).ok()
+                // Step 5
+                .and_then(|decoded_fragid| self.get_element_by_id(&Atom::from(&*decoded_fragid)))
+                // Step 6
+                .or_else(|| self.get_anchor_by_name(fragid))
+                // Step 7
+                .or_else(|| if fragid.to_lowercase() == "top" {
+                    self.GetDocumentElement()
+                } else {
+                    // Step 8
+                    None
+                })
+        }
+    }
+
+    fn get_anchor_by_name(&self, name: &str ) -> Option<Root<Element>> {
+        let check_anchor = |node: &HTMLAnchorElement| {
+            let elem = node.upcast::<Element>();
+            elem.get_attribute(&ns!(), &atom!("name"))
+                .map_or(false, |attr| &**attr.value() == name)
+        };
+        let doc_node = self.upcast::<Node>();
+        doc_node.traverse_preorder()
+                .filter_map(Root::downcast)
+                .find(|node| check_anchor(&node))
```

Comment:
To be honest, this was existing code that I extracted into its own function, and I didn't look too hard at the details.

A quick attempt tells me that this isn't a straightforward change. I don't fully understand the type-conversions going on there, but I can't see this code getting shorter by going down that path.